### ISSUE TITLE
Summary Metric Calculates Percentiles Strangely

### ISSUE THREAD

**Issue by yarmiganosca** [prometheus/client_ruby#76]
Using Ruby 2.4.1, `prometheus-client` 0.7.1, and `quantile` 0.2.0 (for specificity), the Summary class is generating strange percentiles (or what I assume are supposed to be percentiles).
```
➜ bundle exec pry
[1] pry(main)> require 'prometheus/client/summary'
=> true
[2] pry(main)> summary = Prometheus::Client::Summary.new(:a, "a")
=> #<Prometheus::Client::Summary:0x00000003041908
 @base_labels={},
 @docstring="a",
 @mutex=#<Thread::Mutex:0x00000003041890>,
 @name=:a,
 @validator=#<Prometheus::Client::LabelSetValidator:0x00000003041840 @validated={}>,
 @values={}>
[3] pry(main)> (1..100_000).each { |n| summary.observe({}, n) }; summary.get
=> {0.5=>27253, 0.9=>44736, 0.99=>49532}
```
This is something the `quantile` gem is doing, not something `Prometheus::Client::Summary` is doing, as the values returned are identical to the ones provided by the underlying library
```
➜ bundle exec pry
[1] pry(main)> require 'quantile'
=> true
[2] pry(main)> qe = Quantile::Estimator.new
=> #<Quantile::Estimator:0x0000000238aea8
 @buffer=[],
 @head=nil,
 @invariants=
  [#<Quantile::Quantile:0x0000000238ae58 @coefficient_i=0.2, @coefficient_ii=0.2, @inaccuracy=0.05, @quantile=0.5>,
   #<Quantile::Quantile:0x0000000238ae30 @coefficient_i=0.20000000000000004, @coefficient_ii=0.022222222222222223, @inaccuracy=0.01, @quantile=0.9>,
   #<Quantile::Quantile:0x0000000238ae08 @coefficient_i=0.19999999999999982, @coefficient_ii=0.00202020202020202, @inaccuracy=0.001, @quantile=0.99>],
 @observations=0,
 @sum=0>
[3] pry(main)> (1..100_000).each(&qe.method(:observe)); nil
=> nil
[4] pry(main)> qe.query(0.5)
=> 27253
[5] pry(main)> qe.query(0.9)
=> 44736
[6] pry(main)> qe.query(0.99)
=> 49532
```
But it looks like the Java client library has this exact (well, almost) setup as an automated test, and they assert the values are pretty much normal percentiles, but with an error margin:
https://github.com/prometheus/client_java/blob/master/simpleclient/src/test/java/io/prometheus/client/SummaryTest.java#L72-L89
```java
  @Test
  public void testQuantiles() {
    int nSamples = 1000000; // simulate one million samples

    for (int i=1; i<=nSamples; i++) {
      // In this test, we observe the numbers from 1 to nSamples,
      // because that makes it easy to verify if the quantiles are correct.
      labelsAndQuantiles.labels("a").observe(i);
      noLabelsAndQuantiles.observe(i);
    }
    assertEquals(getNoLabelQuantile(0.5), 0.5 * nSamples, 0.05 * nSamples);
    assertEquals(getNoLabelQuantile(0.9), 0.9 * nSamples, 0.01 * nSamples);
    assertEquals(getNoLabelQuantile(0.99), 0.99 * nSamples, 0.001 * nSamples);

    assertEquals(getLabeledQuantile("a", 0.5), 0.5 * nSamples, 0.05 * nSamples);
    assertEquals(getLabeledQuantile("a", 0.9), 0.9 * nSamples, 0.01 * nSamples);
    assertEquals(getLabeledQuantile("a", 0.99), 0.99 * nSamples, 0.001 * nSamples);
  }
```
Those assertions seem to indicate what I thought would happen, which is the percentiles (under uniform distribution) will be linear with the number of observations.

So, which behavior is correct?

**Comment by Bruno-DaSilva** [comment]
Seems to me that the bug is in the quantile gem, and what's in the java client is correct. Seeing as the 50th percentile of 1,000,000 is 500,000 +- error (50,000 in this case), the ruby implementation is WAY off but the java one is still within bounds.

I did find the bug for this, it looks like the gem is double counting some values in some cases. 
I've opened a PR in the gem, any chance you could take a look @grobie? It's a very small change. https://github.com/matttproud/ruby_quantile_estimation/pull/11

Here's the difference in results:
```
numbers from 1 -> 1,000,000
OLD:
50th: 251919
90th: 449900
99th: 495284

NEW:
50th: 523945
90th: 904776
99th: 990792



numbers from 1 -> 100,000 + 50,000 -> 100,000
OLD:
50th: 39616
90th: 62061
99th: 66176

NEW:
50th: 65194
90th: 92842
99th: 99190
```

**Comment by grobie** [comment]
Thanks. Fixed in matttproud/ruby_quantile_estimation#11 which has been released. Going to release a new client gem soon.
